R/expression_convert.R

```diff
@@ -0,0 +1,94 @@
+# Inspiration from Advanced R 
+expr_type <- function(x) {
+  if (rlang::is_syntactic_literal(x)) {
+    "constant"
+  } else if (is.symbol(x)) {
+    "symbol"
+  } else if (is.call(x)) {
+    "call"
+  } else if (is.pairlist(x)) {
+    "pairlist"
+  } else {
+    typeof(x)
+  }
+}
+
+switch_expr <- function(x, ...) {
+  switch(expr_type(x),
+         ...,
+         stop("Don't know how to handle type ", typeof(x), call. = FALSE)
+  )
+}
+
+iteration_removal <- function(x) {
+  switch_expr(
+    x,
+    # Base cases
+    constant = x,
+    symbol = {
+      x
+    },
+    # Recursive cases
+    call = {
+      # go into [ to remove i's or add cbind
+      if (identical(x[[1]], rlang::sym("["))) {
+        # remove the iteration variable i
+        if (identical(x[[3]], rlang::sym("i"))) {
+          return(x[[2]])
+        } 
+        # cbind switch for [ with multiple arguments
+        if (length(x) > 3) {
+          args1 <- purrr::map(as.list(x)[c(3, 4)], iteration_removal)
+          fun1 <- rlang::call2(rlang::expr(cbind), !!!args1)
+          return(rlang::call2(x[[1]], x[[2]], fun1))
+        }
+      }
+      args <- purrr::map(as.list(x)[-1], iteration_removal)
+      rlang::call2(x[[1]], !!!args)
+    },
+    pairlist = {
+      x
+    }
+  )
+}
+
+loop_removal <- function(x) {
+  switch_expr(
+    x,
+    # Base cases
+    constant = x,
+    symbol = {
+      x
+    },
+    # Recursive cases
+    call = {
+      # return body of for loop, ie remove for loop line
+      if (identical(x[[1]], rlang::sym("for"))) {
+        return(x[[4]])
+      }
+      args <- purrr::map(as.list(x)[-1], iteration_removal)
+      rlang::call2(x[[1]], !!!args)
+    },
+    pairlist = {
+      x
+    }
+  )
+}
+
+#' Convert New Expression
+#'
+#' Takes an expression and removes the for loop and adds `cbind` for arrays.
+#'
+#' @param x An expression
+#'
+#' @return An expression
+#' @export
+#'
+#' @examples
+#' expression_convert("log(eCount[i]) <- b0")
+#' expression_convert("for(i in 1:nObs) {eCount[i] <- b0}")
+#' expression_convert("for(i in 1:length(LogLength)) {eWeightLength[i] <- b0 + bDayte * Dayte[i]}")
+#' expression_convert("for(i in 1:nObs) {eAnnual[i] <- bAnn[Ann[i]] + bSA[Site[i], Ann[i]]}")
+expression_convert <- function(x) {
+  loop_removal(iteration_removal(x))
```

Comment:
```suggestion
  loop_removal(x)
```

Can `loop_removal(x)` call `iteration_removal()` and pass the name of the iteration variable to it?

Reply:
I can take a look to see what we can do with that. I think that it should grab the iteration variable in an automated way incase someone passes something other than `i` to it. 

Right now the function works even in the cases when the `new_expr` doesn't have a for loop, these are not common but this case is present within our code base. I am curious for these cases if the user should turn off the functionality of passing the `new_expr` as an expression and use the string version of the `new_expr` as I assume they won't have speed gains from vectorizing the `new_expr` as it is already written that way. @joethorley what are your thoughts on this?

Even if they should turn off this functionality I assume we don't want it to error for people with the case of not having a `for` loop present in the `new_expr`. Thoughts?

Reply:
I added a condition that checks whether a `for` loop is present and if a `for` loop is present the code will remove the for loop and iteration variables and if the for loop isn't present it will just return the expression. 

The main concern with this implementation is if there is code before the `for` loop in the `new_expr` it won't pick it up. I have added a test for this case to demonstrate this behaviour. I did find one example of this case when searching through the various examples.